caffe2/python/examples/cifar10_fp16.py

```diff
@@ -0,0 +1,210 @@
+        fc2 = brew.fc(model, fc1, 'fc2', 64, 10, weight_init=('GaussianFill', {'std' : 0.1}))
+
+    # Final cast out -> fp32
+    if param_type == "float16":
+        fc2 = model.HalfToFloat(fc2, "fc2_fp32")
+
+    softmax = model.Softmax(fc2, 'softmax')
+
+    return softmax
+
+def AddAccuracy(model, softmax, label):
+    """Adds an accuracy op to the model"""
+
+    accuracy = model.Accuracy([softmax, label], "accuracy", top_k=1)
+    return accuracy
+
+def AddTrainingOperators(model, softmax, label, using_snapshot, use_fp16=False):
+    """Adds training operators to the model."""
+    xent = model.LabelCrossEntropy([softmax, label], 'xent')
+    loss = model.AveragedLoss(xent, "loss")
+
+    AddAccuracy(model, softmax, label)
+
+    print("adding gradient ops")
+    model.AddGradientOperators([loss])
+
+
+    ITER = brew.iter(model, "ITER")
+    LR = model.LearningRate(
+        "ITER", "LR", base_lr=0.001, policy="step", stepsize=4000, gamma=0.1)
+    ONE = model.param_init_net.ConstantFill([], "ONE", shape=[1], value=1.0)
+
+    brew.add_weight_decay(model, 0.004)
+
+    for param in model.params:
+        param_grad = model.param_to_grad[param]
+        # Check if we need to use the full precision version of parameters for update
+        if param in model.param_to_float_copy:
+            param_fp32 = model.param_to_float_copy[param]
+            # Initialize the momentum from the full fp32 version of the parameters
+            param_mom = model.param_init_net.ConstantFill([param_fp32], param_grad+'_mom', value=0.)
+
+            # cast fp16 gradients -> fp32 and update
+            grad_fp32 = model.HalfToFloat(param_grad, param_grad+'_fp32')
+            model.MomentumSGDUpdate([grad_fp32, param_mom, LR, param_fp32], [grad_fp32, param_mom, param_fp32],
+                                    momentum=0.9, nesterov=False)
+            # update the fp16 copy of params
+            param = model.FloatToHalf(param_fp32, param)
+        else:
+            param_mom = model.param_init_net.ConstantFill([param], param_grad+'_mom', value=0.)
+            model.MomentumSGDUpdate([param_grad, param_mom, LR, param], [param_grad, param_mom, param], momentum=0.9, nesterov=False)
+
+def AddBookkeepingOperators(model):
+    """This adds a few bookkeeping operators that we can inspect later.
+
+    These operators do not affect the training procedure: they only collect
+    statistics and prints them to file or to logs.
+    """
+    model.Print('accuracy', [], to_file=1)
+    model.Print('loss', [], to_file=1)
+    for param in model.params:
+        model.Summarize(param, [], to_file=1)
+        model.Summarize(model.param_to_grad[param], [], to_file=1)
+
+def createTrainModel(param_type="float"):
+    order_scope = { 'order' : 'NCHW' }
+    train_model = ModelHelper(name="cifar10_quick_train", arg_scope=order_scope)
+
+    with core.DeviceScope(core.DeviceOption(caffe2_pb2.CUDA, 0)):
```

Comment:
This is explicitly a GPU test. AFAIK CPU operators don't have any fp16 implementations.